packages/mui-base/src/Popup/Popup.tsx

```diff
@@ -0,0 +1,294 @@
+'use client';
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import { autoUpdate, flip, offset, useFloating, VirtualElement } from '@floating-ui/react-dom';
+import {
+  HTMLElementType,
+  unstable_useEnhancedEffect as useEnhancedEffect,
+  unstable_useForkRef as useForkRef,
+} from '@mui/utils';
+import composeClasses from '../composeClasses';
+import Portal from '../Portal';
+import { useSlotProps } from '../utils';
+import { useClassNamesOverride } from '../utils/ClassNameConfigurator';
+import { getPopupUtilityClass } from './popupClasses';
+import { PopupChildrenProps, PopupOwnerState, PopupProps } from './Popup.types';
+
+function useUtilityClasses(ownerState: PopupOwnerState) {
+  const { open } = ownerState;
+
+  const slots = {
+    root: ['root', open && 'open'],
+  };
+
+  return composeClasses(slots, useClassNamesOverride(getPopupUtilityClass));
+}
+
+function resolveAnchor(
+  anchor:
+    | VirtualElement
+    | (() => VirtualElement)
+    | HTMLElement
+    | (() => HTMLElement)
+    | null
+    | undefined,
+): HTMLElement | VirtualElement | null | undefined {
+  return typeof anchor === 'function' ? anchor() : anchor;
+}
+/**
+ *
+ * Demos:
+ *
+ * - [Popup](https://mui.com/base-ui/react-popup/)
+ *
+ * API:
+ *
+ * - [Popup API](https://mui.com/base-ui/react-popup/components-api/#popup)
+ */
+const Popup = React.forwardRef(function Popup<RootComponentType extends React.ElementType>(
+  props: PopupProps<RootComponentType>,
+  forwardedRef: React.ForwardedRef<Element>,
+) {
+  const {
+    anchor: anchorProp,
+    children,
+    container,
+    disablePortal = false,
+    keepMounted = false,
+    middleware,
+    offset: offsetProp = 0,
+    open = false,
+    placement = 'bottom',
+    slotProps = {},
+    slots = {},
+    strategy = 'absolute',
+    withTransition = false,
+    ...other
+  } = props;
+
+  const {
+    refs,
+    elements,
+    floatingStyles,
+    update,
+    placement: finalPlacement,
+  } = useFloating({
```

Comment:
The usage of the hook looks great, I couldn't find anything to complain about :) 👍